adr/ADR-48.md

```diff
@@ -0,0 +1,137 @@
+
+Related:
+
+ * [Netflix Distributed Counter](https://netflixtechblog.com/netflixs-distributed-counter-abstraction-8d0c45eb66b2)
+
+## Solution Overview
+
+A Stream can opt-in to supporting Counters which will allow any subject to be a counter.
+
+Publishing a message to such a Stream will load the most recent message on the subject, increment its value and save 
+the new message with the latest value in the body. The header is preserved for downstream processing by Sources and for 
+visibility and debugging purposes.
+
+```bash
+$ nats s get COUNTER --last-for counter.hits
+Item: COUNTER#22802062 received 2025-01-09 18:05:07.93747413 +0000 UTC on Subject counter.hits
+
+Headers:
+  Nats-Incr: +2
+
+{"val":"100"}
+$ nats pub counter.hits '' -J -H "Nats-Incr:+1"
+$ nats s get COUNTER --last-for counter.hits
+Item: COUNTER#22802063 received 2025-01-09 18:06:00 +0000 UTC on Subject counter.hits
+
+Headers:
+  Nats-Incr: +1
+
+{"val":"101"}
+```
+
+## Design and Behavior
+
+The goal is to support addition and subtraction only and that a Stream that Sources 10s of other Streams all with 
+counters will effectively create a big combined counter holding totals contributed to by all sources.
+
+Handling published messages has the follow behavior and constraints:
+
+ * The header holds values like `+1`, `-1` and `-10`, in other words any valid `int64`, if the value fails to parse 
+   the message is rejected with an error. A value of `0` is valid.
+ * When publishing a message to the subject the last value is loaded by the server receiving the message (so the leader), the body is parsed, incremented and written 
+   into the new message body. The headers are all preserved.
+ * If the addition will overflow a `int64` in either direction the message will be rejected with an error
+ * When publishing a message and the previous message do not have a `Nats-Incr` header it means the subject is not a 
+   counter, an error is returned to the user and the message is rejected
+ * When a message has a `Nats-Rollup`, `Nats-Expected-Last-Sequence` or `Nats-Expected-Subject-Last-Sequence`, `Nats-Expected-Stream` or `Nats-Expected-Last-Msg-Id` header in addition to a `Nats-Incr` header it should be rejected
+ * When a message with the header is received over a Source, that has the configuration setting enabled, processing is 
+   done as above otherwise the message is discarded
+ * When a message with the header is published to a Stream without the option set the message is rejected with an error
+ * When a message with the header is received over a Mirror the message is stored verbatim
+ * When a message without the header is received and the previous message is a counter this will be rejected since 
+   the subject is then treated as being a counter only. This could be very expensive on Streams with many 
+   non-counter messages being written to them
+
+The value in the body is stored in a struct with the following layout:
+
+```go
+type CounterValue struct {
```

Comment:
Not convinced that wrapping in JSON is better here than just, say, encoding as a varint to start with. Varint avoids a whole `atoi`/`atof` mess with strings and it avoids the 2^53 limitation of numbers in JSON. It could even be extended beyond 64-bit later if needed.

Even if we need to add more fields later, we can just add anything else we need (or even a JSON blob if it came to it) after the varint.

Reply:
Too many NATS clients to support that may or may not do var int IMO. I think the value as a string and the client libs parse as they see fit makes sense for now.

Reply:
FWIW at the rate Netflix states in their blog post a 2^52 will last over 2000 years.

Reply:
So are you making the case to make it an int value?

Reply:
+1 for 2^52 
Java has a signed 64 bit integer and a signed 64 bit float native. It does have object versions for bigger numbers.

Reply:
Just adding in color. We might be over thinking it. These are very big numbers. 

But also of course in modern times and internet scale perhaps not that big. 

More realistically we can look at our theoretical messages rate of let’s say 10k/sec as a target and then really 10^52 is a very very big number. (14k years) assuming +1 counts, but that might be naive to assume small increments. 

So it’s hard to say. I am inclined to do strong limited to 2^53 for now. It keeps it easy without closing the door.

Reply:
I agree with a string. It's also the safest approach against weird way different languages deserialize non-string values.